File: original/match.c
#include "copyright.h"

/* Routines for parsing arguments */
#include <stdlib.h>
#include <ctype.h>

#include "db.h"
#include "config.h"
#include "externs.h"
#include "interface.h"
#include "match.h"

#define DOWNCASE(x) (isupper(x) ? tolower(x) : (x))

static dbref exact_match = NOTHING;	/* holds result of exact match */
static int check_keys = 0;	/* if non-zero, check for keys */
static dbref last_match = NOTHING;	/* holds result of last match */
static int match_count;		/* holds total number of inexact matches */
static dbref match_who;	/* player who is being matched around */
static const char *match_name;	/* name to match */
static int preferred_type = NOTYPE; /* preferred type */

void init_match(dbref player, const char *name, int type)
{
    exact_match = last_match = NOTHING;
    match_count = 0;
    match_who = player;
    match_name = name;
    check_keys = 0;
    preferred_type = type;
}

void init_match_check_keys(dbref player, const char *name, int type)
{
    init_match(player, name, type);
    check_keys = 1;
}

static dbref choose_thing(dbref thing1, dbref thing2)
{
    int has1;
    int has2;

    if(thing1 == NOTHING) {
	return thing2;
    } else if(thing2 == NOTHING) {
	return thing1;
    }

    if(preferred_type != NOTYPE) {
	if(Typeof(thing1) == preferred_type) {
	    if(Typeof(thing2) != preferred_type) {
		return thing1;
	    }
	} else if(Typeof(thing2) == preferred_type) {
	    return thing2;
	}
    }

    if(check_keys) {
	has1 = could_doit(match_who, thing1);
	has2 = could_doit(match_who, thing2);

	if(has1 && !has2) {
	    return thing1;
	} else if (has2 && !has1) {
	    return thing2;
	}
	/* else fall through */
    }

    return (random() % 2 ? thing1 : thing2);
}

void match_player()
{
    dbref match;
    const char *p;

    if(*match_name == LOOKUP_TOKEN
       && payfor(match_who, LOOKUP_COST)) {
	for(p = match_name + 1; isspace(*p); p++);
	if((match = lookup_player(p)) != NOTHING) {
	    exact_match = match;
	}
    }
}

/* returns nnn if name = #nnn, else NOTHING */
static dbref absolute_name()
{
    dbref match;

    if(*match_name == NUMBER_TOKEN) {
	match = parse_dbref(match_name+1);
	if(match < 0 || match >= db_top) {
	    return NOTHING;
	} else {
	    return match;
	}
    } else {
	return NOTHING;
    }
}

void match_absolute()
{
    dbref match;

    if((match = absolute_name()) != NOTHING) {
	exact_match = match;
    }
}

void match_me()
{
    if(!string_compare(match_name, "me")) {
	exact_match = match_who;
    }
}

void match_here()
{
    if(!string_compare(match_name, "here")
       && db[match_who].location != NOTHING) {
	exact_match = db[match_who].location;
    }
}

static void match_list(dbref first)
{
    dbref absolute;

    absolute = absolute_name();
    if(!controls(match_who, absolute)) absolute = NOTHING;

    DOLIST(first, first) {
	if(first == absolute) {
	    exact_match = first;
	    return;
	} else if(!string_compare(db[first].name, match_name)) {
	    /* if there are multiple exact matches, randomly choose one */
	    exact_match = choose_thing(exact_match, first);
	} else if(string_match(db[first].name, match_name)) {
	    last_match = first;
	    match_count++;
	}
    }
}
    
void match_possession()
{
    match_list(db[match_who].contents);
}

void match_neighbor()
{
    dbref loc;

    if((loc = db[match_who].location) != NOTHING) {
	match_list(db[loc].contents);
    }
}

void match_exit()
{
    dbref loc;
    dbref exit;
    int exit_status;
    dbref absolute;
    const char *match;
    const char *p;

    if((loc = db[match_who].location) != NOTHING) {
	absolute = absolute_name();
	if(!controls(match_who, absolute)) absolute = NOTHING;

	DOLIST(exit, db[loc].exits) {
	    if(exit == absolute) {
		exact_match = exit;
	    } else {
		match = db[exit].name;
		while(*match) {
		    /* check out this one */
		    for(p = match_name;
			(*p
			 && DOWNCASE(*p) == DOWNCASE(*match)
			 && *match != EXIT_DELIMITER);
			p++, match++);
		    /* did we get it? */
		    if(*p == '\0') {
			/* make sure there's nothing afterwards */
			while(isspace(*match)) match++;
			if(check_keys) {
			    exit_status = could_doit(match_who, exit);
			} else {
			    exit_status = 1;
			}
			if(*match == '\0' || *match == EXIT_DELIMITER) {
			    /* we got it */
			    exact_match = choose_thing(exact_match, exit);
			    goto next_exit;	/* got this match */
			}
		    }
		    /* we didn't get it, find next match */
		    while(*match && *match++ != EXIT_DELIMITER);
		    while(isspace(*match)) match++;
		}
	    }
	  next_exit:
	    ;
	}
    }
    // Stop GCC complaining - Note there is a bug related to this which we have
    // fixed in the code being ported, see, ext/match.c
    exit_status = exit_status;
}

void match_everything()
{
    match_exit();
    match_neighbor();
    match_possession();
    match_me();
    match_here();
    if(Wizard(match_who)) {
	match_absolute();
	match_player();
    }
}

dbref match_result()
{
    if(exact_match != NOTHING) {
	return exact_match;
    } else {
	switch(match_count) {
	  case 0:
	    return NOTHING;
	  case 1:
	    return last_match;
	  default:
	    return AMBIGUOUS;
	}
    }
}
	   
/* use this if you don't care about ambiguity */
dbref last_match_result()
{
    if(exact_match != NOTHING) {
	return exact_match;
    } else {
	return last_match;
    }
}

dbref noisy_match_result()
{
    dbref match;

    switch(match = match_result()) {
      case NOTHING:
	notify(match_who, NOMATCH_MESSAGE);
	return NOTHING;
      case AMBIGUOUS:
	notify(match_who, AMBIGUOUS_MESSAGE);
	return NOTHING;
      default:
	return match;
    }
}

File: lib/match.rb
require_relative 'helpers'

module MangledMud
  
  # Handles finding (matching) "things" through searching a number of different "locations"
  # @note This follows the original code to the letter, further refactoring would make this more ruby-esque
  # @note This documentation is an overview, the code is subtle and requires inspection for nuances
  #
  # @version 1.0
  class Match
    include Helpers

    # @param [Db] db the current database instance
    # @param [Object] notifier An object with method do_notify(player_id, string), the method will be called to send notifications back to the player invoking the command/action
    def initialize(db, notifier)
      @db = db
      @notifier = notifier
    end

    # Initialize, ready for a new match to be made, match ignoring access rights for the thing being matched.
    # @note Must be called before any of the match_ methods
    #
    # @param [Number] player the database record number for the player
    # @param [String] name the name of the thing to be matched, may be nil
    # @param [Number] type the type of the thing to be matched, see {TYPE_ROOM}, {TYPE_THING}, {TYPE_EXIT}, {TYPE_PLAYER}, {NOTYPE}
    def init_match(player, name, type)
      @exact_match = NOTHING
      @last_match = NOTHING
      @match_count = 0
      @match_who = player
      @match_name = name
      @check_keys = false
      @preferred_type = type
    end

    # Initialize, ready for a new match to be made, match taking into account the players access rights for the thing being matched
    # @note Must be called before any of the match_ methods
    #
    # @param [Number] player the database record number for the player
    # @param [String] name the name of the thing to be matched, may be nil
    # @param [Number] type the type of the thing to be matched, see {TYPE_ROOM}, {TYPE_THING}, {TYPE_EXIT}, {TYPE_PLAYER}, {NOTYPE}
    def init_match_check_keys(player, name, type)
      init_match(player, name, type)
      @check_keys = true
    end

    # Match a player by name (as set in {#init_match_check_keys} or {#init_match})
    # Only an exact match is possible.
    #
    # @see #match_result
    # @see #last_match_result
    # @see #noisy_match_result
    def match_player()
      if (@match_name and @match_name[0] == LOOKUP_TOKEN && Predicates.new(@db, @notifier).payfor(@match_who, LOOKUP_COST))
        player_id = @match_name[1..-1].lstrip()
        match = Player.new(@db, @notifier).lookup_player(player_id)
        @exact_match = match if (match != NOTHING)
      end
    end

    # Match by absolute name, i.e. name is #nnn, where nnn is a database record identifier (number)
    # Only an exact match is possible. See {#init_match_check_keys} or {#init_match}.
    #
    # @see #match_result
    # @see #last_match_result
    # @see #noisy_match_result
    def match_absolute()
      match = absolute_name()
      @exact_match = match if (match != NOTHING)
    end

    # If matching on "me" then exact match on the player
    # Only an exact match is possible. See {#init_match_check_keys} or {#init_match}.
    def match_me()
      if @match_name
        @exact_match = @match_who if (@match_name.casecmp("me") == 0)
      end
    end

    # If matching on "here" then exact match on the players location
    # Only an exact match is possible. See {#init_match_check_keys} or {#init_match}.
    def match_here()
      if (@match_name and @match_name.casecmp("here") == 0 && @db[@match_who].location != NOTHING)
        @exact_match = @db[@match_who].location
      end
    end

    # Match a thing in players contents, an ambiguous match may occur
    # (if the thing is not specified by an absolute name).
    # See {#init_match_check_keys} or {#init_match}.
    #
    # @see #match_result
    # @see #last_match_result
    # @see #noisy_match_result
    def match_possession()
      match_list(@db[@match_who].contents)
    end

    # Match the player based on the contents of their location, an ambiguous match may occur
    # (if the thing is not specified by an absolute name)
    # See {#init_match_check_keys} or {#init_match}.
    #
    # @see #match_result
    # @see #last_match_result
    # @see #noisy_match_result
    def match_neighbor()
      loc = @db[@match_who].location
      if (loc != NOTHING)
        match_list(@db[loc].contents)
      end
    end

    # Match an exit based on the location of the player, an ambiguous match may occur
    # (if the exit is not specified by an absolute name)
    # See {#init_match_check_keys} or {#init_match}.
    #
    # @see #match_result
    # @see #last_match_result
    # @see #noisy_match_result
    def match_exit()
      loc = @db[@match_who].location
      if (loc != NOTHING)
        absolute = absolute_name()
        absolute = NOTHING if (!Predicates.new(@db, @notifier).controls(@match_who, absolute))

        enum(@db[loc].exits).each do |exit|
          if (exit == absolute)
            @exact_match = exit
          elsif @match_name
            @db[exit].name.split(EXIT_DELIMITER).each do |name|
              # Allow a partial match - for ambiguous matching
              if (name.downcase.strip.start_with?(@match_name.downcase))
                # ! Matthew - Modified original code -> Bug fix?
                if (@check_keys)
                  could_doit = Predicates.new(@db, @notifier).could_doit(@match_who, exit)
                  @match_count += 1 if could_doit
                else
                  @match_count += 1
                end
                # Only match exact if the names are equal
                if (@match_name.casecmp(name) == 0)
                  @exact_match = choose_thing(@exact_match, exit)
                end
              end
            end
          end
        end
      end
    end

    # Perform {#match_exit}, {#match_neighbor}, {#match_possession}, {#match_me} and {#match_here},
    # if the player is a wizard then also {#match_absolute} and {#match_player}
    def match_everything()
      match_exit()
      match_neighbor()
      match_possession()
      match_me()
      match_here()
      if (is_wizard(@match_who))
        match_absolute()
        match_player()
      end
    end

    # Get the result of calling a match command
    #
    # @return [Number] if an exact match occurred, then return the identifier of the database record that matched otherwise, return {NOTHING} or {AMBIGUOUS}
    def match_result()
      if (@exact_match != NOTHING)
        @exact_match
      else
        case @match_count
        when 0 then NOTHING
        when 1 then @last_match
        else AMBIGUOUS
        end
      end
    end

    # Return the last item matched
    # @return [Number] if a match occurred, return the identifier of the database record that matched else {NOTHING}
    def last_match_result()
      (@exact_match != NOTHING) ? @exact_match : @last_match
    end

    # The same as {#match_result}, except this uses the notifier passed to the initializer to inform the player of the match result.
    # @return [Number] if an exact match occurred, then return the identifier of the database record that matched otherwise, return {NOTHING} or {AMBIGUOUS}
    def noisy_match_result()
      match = match_result()
      case match
      when NOTHING
        @notifier.do_notify(@match_who, Phrasebook.lookup('dont-see-that'))
        NOTHING
      when AMBIGUOUS
        @notifier.do_notify(@match_who, Phrasebook.lookup('which-one'))
        NOTHING
      else
        match
      end
    end

    private

    # returns nnn if name = #nnn, else NOTHING
    def absolute_name()
      if (@match_name and @match_name[0] == NUMBER_TOKEN)
        match = @db.parse_dbref(@match_name[1..-1])
        if (match < 0 || match >= @db.length)
          return NOTHING
        else
          match
        end
      else
        NOTHING
      end
    end

    # Enumerate a list starting at first, looking for the thing to match
    def match_list(first)
      absolute = absolute_name()
      absolute = NOTHING if (!Predicates.new(@db, @notifier).controls(@match_who, absolute))

      enum(first).each do |i|
        if (i == absolute)
          @exact_match = i
          return
        elsif @match_name
          if (@db[i].name.casecmp(@match_name) == 0)
            # if there are multiple exact matches, randomly choose one
            @exact_match = choose_thing(@exact_match, i)
          else
            # Match at the start of words - There must be a neater way
            name_words = @db[i].name.split(/\s+/)
            name_words.each do |word|
              if word.downcase.start_with?(@match_name.downcase)
                @last_match = i
                @match_count += 1
                break
              end
            end
          end
        end
      end
    end

    # Given a choice of two things, pick one of them (one or both things may be NOTHING)
    def choose_thing(thing1, thing2)
      if (thing1 == NOTHING)
        return thing2
      elsif (thing2 == NOTHING)
        return thing1
      end

      if (@preferred_type != NOTYPE)
        if (typeof(thing1) == @preferred_type)
          if (typeof(thing2) != @preferred_type)
            return thing1
          end
        elsif (typeof(thing2) == @preferred_type)
          return thing2
        end
      end

      if (@check_keys)
        has1 = Predicates.new(@db, @notifier).could_doit(@match_who, thing1)
        has2 = Predicates.new(@db, @notifier).could_doit(@match_who, thing2)
        if (has1 && !has2)
          return thing1
        elsif (has2 && !has1)
          return thing2
        end
        # else fall through
      end

      return (Game.do_rand() % 2 ? thing1 : thing2)
    end
  end
end
